File: src/lib/desktop/projects.ts
import { invoke } from '@tauri-apps/api/core';
import type { TauriActionResult } from '@/lib/desktop/admin-types';
import { isTauriBackendAvailable } from '@/lib/desktop/runtime';

function requireTauri(): void {
  if (!isTauriBackendAvailable()) {
    throw new Error('Tauri projects API requires the desktop shell');
  }
}

export async function tauriProjectsGetOrgMembersForTasks() {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_org_members_for_tasks_cmd');
}

export async function tauriProjectsCreateProject(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_create_project_cmd', { data });
}

export async function tauriProjectsGetProject(projectId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_get_project_cmd', { projectId });
}

export async function tauriProjectsGetProjectsByOrg(orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_projects_by_org_cmd', { orgId });
}

export async function tauriProjectsGetProjectsForCurrentOrg(limit?: number) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_projects_for_current_org_cmd', { limit: limit ?? null });
}

export async function tauriProjectsGetProjectsByStatus(input: { orgId: string; status: string }) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_projects_by_status_cmd', { input });
}

export async function tauriProjectsUpdateProject(projectId: string, data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_update_project_cmd', { projectId, data });
}

export async function tauriProjectsAddMemberToProject(input: { projectId: string; userId: string }) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_add_member_to_project_cmd', { input });
}

export async function tauriProjectsDeleteProject(projectId: string) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('projects_delete_project_cmd', { projectId });
}

export async function tauriProjectsCreateTask(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_create_task_cmd', { data });
}

export async function tauriProjectsGetTask(taskId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_get_task_cmd', { taskId });
}

export async function tauriProjectsGetTasksByProject(projectId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_tasks_by_project_cmd', { projectId });
}

export async function tauriProjectsGetTasksByOrg(orgId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_tasks_by_org_cmd', { orgId });
}

export async function tauriProjectsGetTasksForCurrentOrg(limit?: number) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_tasks_for_current_org_cmd', { limit: limit ?? null });
}

export async function tauriProjectsGetTasksByAssigned(userId: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_tasks_by_assigned_cmd', { userId });
}

export async function tauriProjectsGetTasksByStatus(input: { orgId: string; status: string }) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('projects_get_tasks_by_status_cmd', { input });
}

export async function tauriProjectsUpdateTask(taskId: string, data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_update_task_cmd', { taskId, data });
}

export async function tauriProjectsDeleteTask(taskId: string) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('projects_delete_task_cmd', { taskId });
}

export async function tauriProjectsGetWorkBoardConfig() {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_get_work_board_config_cmd');
}

export async function tauriProjectsSaveWorkBoardConfig(data: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_save_work_board_config_cmd', { data });
}

export async function tauriProjectsMigrateTaskStatuses(input: { fromStatus: string; toStatus: string }) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_migrate_task_statuses_cmd', { input });
}

export async function tauriProjectsMigrateProjectStatuses(input: { fromStatus: string; toStatus: string }) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('projects_migrate_project_statuses_cmd', { input });
}

File: src-tauri/src/commands/projects.rs
use serde::Deserialize;
use serde_json::Value;
use tauri::State;

use crate::admin::service::ActionResult;
use crate::auth::session::SessionState;
use crate::projects::service::{
    add_member_to_project, create_project, create_task, delete_project, delete_task, get_org_members_for_tasks,
    get_project, get_projects_by_org, get_projects_by_status, get_projects_for_current_org, get_task,
    get_tasks_by_assigned, get_tasks_by_org, get_tasks_by_project, get_tasks_by_status, get_tasks_for_current_org,
    get_work_board_config, migrate_project_statuses, migrate_task_statuses, save_work_board_config, update_project,
    update_task,
};
use crate::state::AppState;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectsByStatusInput {
    pub org_id: String,
    pub status: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddMemberToProjectInput {
    pub project_id: String,
    pub user_id: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TasksByStatusInput {
    pub org_id: String,
    pub status: String,
}

#[tauri::command]
pub fn projects_get_org_members_for_tasks_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Vec<Value>> {
    get_org_members_for_tasks(&app, &session)
}

#[tauri::command]
pub fn projects_create_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    create_project(&app, &session, data)
}

#[tauri::command]
pub fn projects_get_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    project_id: String,
) -> ActionResult<Value> {
    get_project(&app, &session, &project_id)
}

#[tauri::command]
pub fn projects_get_projects_by_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
) -> ActionResult<Vec<Value>> {
    get_projects_by_org(&app, &session, &org_id)
}

#[tauri::command]
pub fn projects_get_projects_for_current_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_projects_for_current_org(&app, &session, limit.unwrap_or(200))
}

#[tauri::command]
pub fn projects_get_projects_by_status_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    input: ProjectsByStatusInput,
) -> ActionResult<Vec<Value>> {
    get_projects_by_status(&app, &session, &input.org_id, &input.status)
}

#[tauri::command]
pub fn projects_update_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    project_id: String,
    data: Value,
) -> ActionResult<Value> {
    update_project(&app, &session, &project_id, data)
}

#[tauri::command]
pub fn projects_add_member_to_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    input: AddMemberToProjectInput,
) -> ActionResult<Value> {
    add_member_to_project(&app, &session, &input.project_id, &input.user_id)
}

#[tauri::command]
pub fn projects_delete_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    project_id: String,
) -> ActionResult<bool> {
    delete_project(&app, &session, &project_id)
}

#[tauri::command]
pub fn projects_create_task_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    create_task(&app, &session, data)
}

#[tauri::command]
pub fn projects_get_task_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    task_id: String,
) -> ActionResult<Value> {
    get_task(&app, &session, &task_id)
}

#[tauri::command]
pub fn projects_get_tasks_by_project_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    project_id: String,
) -> ActionResult<Vec<Value>> {
    get_tasks_by_project(&app, &session, &project_id)
}

#[tauri::command]
pub fn projects_get_tasks_by_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
) -> ActionResult<Vec<Value>> {
    get_tasks_by_org(&app, &session, &org_id)
}

#[tauri::command]
pub fn projects_get_tasks_for_current_org_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_tasks_for_current_org(&app, &session, limit.unwrap_or(300))
}

#[tauri::command]
pub fn projects_get_tasks_by_assigned_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    user_id: String,
) -> ActionResult<Vec<Value>> {
    get_tasks_by_assigned(&app, &session, &user_id)
}

#[tauri::command]
pub fn projects_get_tasks_by_status_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    input: TasksByStatusInput,
) -> ActionResult<Vec<Value>> {
    get_tasks_by_status(&app, &session, &input.org_id, &input.status)
}

#[tauri::command]
pub fn projects_update_task_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    task_id: String,
    data: Value,
) -> ActionResult<Value> {
    update_task(&app, &session, &task_id, data)
}

#[tauri::command]
pub fn projects_delete_task_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    task_id: String,
) -> ActionResult<bool> {
    delete_task(&app, &session, &task_id)
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusMigrationInput {
    pub from_status: String,
    pub to_status: String,
}

#[tauri::command]
pub fn projects_get_work_board_config_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    get_work_board_config(&app, &session)
}

#[tauri::command]
pub fn projects_save_work_board_config_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    data: Value,
) -> ActionResult<Value> {
    save_work_board_config(&app, &session, data)
}

#[tauri::command]
pub fn projects_migrate_task_statuses_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    input: StatusMigrationInput,
) -> ActionResult<Value> {
    migrate_task_statuses(&app, &session, &input.from_status, &input.to_status)
}

#[tauri::command]
pub fn projects_migrate_project_statuses_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    input: StatusMigrationInput,
) -> ActionResult<Value> {
    migrate_project_statuses(&app, &session, &input.from_status, &input.to_status)
}
